### PR TITLE
3.7 Improve Cache Controller

### PR DESCRIPTION
**Pull request by csthomas**
Pull Request for Issue #12133 and other optimizations.

### Summary of Changes
- Deprecate `get` and `store` methods in `JCacheController`, subclasses should have own methods.
- Deprecate `start` and `end` methods in `JCacheControllerOutput` - joomla does not use it
- Deprecate `call` method in `JCacheControllerCallback`
- Remove `setLifeTime`, `setCaching` from JCacheController because `__call()` method can call it directly on JCache instance.
- Replace operator == to === for `locked` and `locklooped` variables.
- Do not store if lock can not be locked:

``` php
if ($locktest->locked === false && $locktest->locklooped === true)
```
- Add `get` and `store` method to `JCacheControllerOutput` which is the same as deprecated method `get` and `store` in `JCacheController`. 
- In subclasses of `JCacheController` I did optimization on `get` methods.
  - move `unlock()` method before unserialize(...)
  - Generally unlock cache as fast as we can
- Do not use ternary operator on variables with a lots of data, see http://fabien.potencier.org/the-php-ternary-operator-fast-or-not.html

### Testing Instructions
Joomla should work as before.
Code review.

### Documentation Changes Required
None.

### CODE REVIEW

**Comment by csthomas** [comment]
Can I ask you @mbabker for a review?

File: libraries/joomla/cache/controller/callback.php [hunk 1]
```diff
@@ -88,86 +89,102 @@ public function get($callback, $args = array(), $id = false, $wrkarounds = false
...
+		$result = call_user_func_array($callback, $referenceArgs);
+		$output = ob_get_clean();
+
+		$data = array('result' => $result);
+
+		if ($wrkarounds)
+		{
+			$data['output'] = JCache::setWorkarounds($output, $coptions);
+		}
+		else {
```
Comment (mbabker) [inline libraries/joomla/cache/controller/callback.php]: PHPCS issue.

**Review by mbabker — CHANGES_REQUESTED** [review]
Looks fine, one PHPCS issue.

**Review by mbabker — APPROVED** [review]

**Comment by csthomas** [comment]
Thanks Michael.

Now I have to wait for 2 successful tests:)

**Comment by franz-wohlkoenig** [comment]
@csthomas Enable Cache and test if J works as before?

**Comment by csthomas** [comment]
Yes, be sure that cache works before patch

**Comment by franz-wohlkoenig** [comment]
I have tested this item :white_check_mark: successfully on 87f42d59c69e092b8b45d5bd0589fa5116a4c90e<br /><br />Tested on Articles, Blog, List, Contact, Feed, Search.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12312">issues.joomla.org/tracker/joomla-cms/12312</a>.</sub>

**Comment by JoshuaLewis** [comment]
I have tested this item :white_check_mark: successfully on 87f42d59c69e092b8b45d5bd0589fa5116a4c90e<br /><br />Logged in, logged out, created article, viewed list, viewed article, changed site settings on frontend, ect. Seems to work great on every view I tested. Successfully tested this PR.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12312">issues.joomla.org/tracker/joomla-cms/12312</a>.</sub>

**Comment by jeckodevelopment** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/12312">issues.joomla.org/tracker/joomla-cms/12312</a>.</sub>

**Pull request merged**